### DISCUSSION TITLE
Fazer um POST enviando Certificado Digital

### DISCUSSION THREAD

**Discussion by marcelinoap** [viniciussanchez/RESTRequest4Delphi#195] [Q&A] [answered]
Estou consultando um serviço da SERPRO e preciso fazer o post enviando o certificado da empresa... consigo fazer pelo POSTMAN pois dá para selecionar o certificado digital A1 por lá. 
Tem como fazer isso pelo TRequest?

```
        Resp := TRequest.New.BaseURL('https://autenticacao.sapi.serpro.gov.br/authenticate')
                        .AddHeader('Authorization', 'Basic '+Chave)
                        .ContentType('application/x-www-form-urlencoded')
                        .Post;
```

**Comment by viniciussanchez** [comment, marked as answer]
Sim.. utilizando a engine (provider) do Indy
https://github.com/viniciussanchez/RESTRequest4Delphi/pull/110

**Reply by marcelinoap** [reply]
Valeu! Muito obrigado! Seu trabalho facilitou muito o nosso!

**Reply by josedachao** [reply]
Olá, a todos.
Fiquei bastante interessado e animado ao saber que poderia fazer isso com o RR4D, mas após bater um pouco de cabeça e admitir que afinal não entendi o que o Vinícius disse ... resolvi pedir um pouco de ajuda por aqui ... Como faria isso (inserir o certificado na requisição) com a engine do Indy?

**Reply by viniciussanchez** [reply]
Primeiramente você precisa mudar a engine do RR4D, colocando a diretiva de compilação referente ao Indy... depois só usar as propriedades CertFile e KeyFile da requisição para informar os dados...

**Reply by marcelinoap** [reply]
Você precisa criar os dois arquivos (CertFile e KeyFile) com base no certificado A1. Ex.: No prompt do DOS
openssl pkcs12 -in <filename.pfx> -clcerts -nokeys -out certificate.crt
openssl pkcs12 -in <filename.pfx> -nocerts -nodes -out private.key


Depois é só adicionar a opção de compilação no seu projeto para que o RR4D compile com estas propriedades do Indy. (Dê uma olhada no readme do diretório do componente)
![image](https://github.com/viniciussanchez/RESTRequest4Delphi/assets/35923635/8f2203aa-34c2-467f-aef4-aa057e9820f2)

**Reply by marcelinoap** [reply]
Dê um "BUILD" no projeto para que as propriedades apareçam.
O meu ficou assim:

```
        Resp := TRequest.New.BaseURL('https://autenticacao.sapi.serpro.gov.br/authenticate')
                        .AddHeader('User-Agent','Goaccount')
                        .AddHeader('Accept','*/*')
                        .AddHeader('Accept-Encoding','gzip, deflate, br')
                        .AddHeader('Authorization', 'Basic '+Chave)
                        .AddHeader('role-type','TERCEIROS')
                        .ContentType('application/x-www-form-urlencoded')
                        .CertFile(PathCrt)
                        .KeyFile(PathKey)
                        .Post;
```
